### PR TITLE
Don't try to modify dict while looping through it

### PR DESCRIPTION
**Pull request by techhat**
### What does this PR do?
As it says.

### What issues does this PR fix or reference?
None known.

### Previous Behavior
You would get an error about trying to modify a dictionary while looping through it.

### New Behavior
No more error.

### Tests written?
No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

### CODE REVIEW

**Review by rallytime — APPROVED** [review]

**Comment by gtmanfred** [comment]
So, this used to be `providers.keys()`, then .keys() got removed because of the purge that pedro did.

So i went back, and this should generate it's own list on the start of the for loop, then that list is seperate from the dictionary, so you can change the dictionary, but the keys list will remain the same?

I am not sure why we would need to copy it.

```
$ cat this.py
this = {'one': 1, 'two': 2, 'three': 3}
for key in list(this):
    print(this.pop(key))
print(this)
$ python this.py
1
2
3
{}
```

**Comment by techhat** [comment]
I don't know why either, but I can tell you that that I was getting errors about modifying the dictionary while looping through it, and the `copy()` stops that.

**Comment by damon-atkins** [comment]
Is this a python 3 issue?
```
Python 3.4.3 (default, Aug  9 2016, 15:44:30)
[GCC 5.3.1 20160406 (Red Hat 5.3.1-6)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> a={1: 2 , 3: 4 }
>>> type(a)
<class 'dict'>
>>> type(a.keys())
<class 'dict_keys'>
>>> type(list(a))
<class 'list'>
```

```
Python 2.7.11 (default, Sep 29 2016, 13:49:11)
[GCC 5.3.1 20160406 (Red Hat 5.3.1-6)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> a={1: 2 , 3: 4 }
>>> type(a.keys())
<type 'list'>
```

**Comment by gtmanfred** [comment]
@damon-atkins that is why we switched to using list() instead, so that it would instantiate a list in both cases.

**Review by gtmanfred — APPROVED** [review]

**Comment by techhat** [comment]
Actually, I think I was in Python 3 when I got this error.

**Comment by gtmanfred** [comment]
Yeah, my default python is 3.6.2, 

```
$ python --version
Python 3.6.2
```

It shouldn't be an issue, but I don't have a good reason not to merge it if you have still seen a problem in some instances.

File: salt/cloud/__init__.py [hunk 1]
```diff
@@ -234,7 +234,7 @@ def _opts_defaults(self, **kwargs):
                          if a.get('provider', '')]
             if providers:
                 _providers = opts.get('providers', {})
-                for provider in list(_providers):
+                for provider in list(_providers).copy():
```
Comment (cachedout) [inline salt/cloud/__init__.py]: I'm not sure about the entirety of this data structure but are you sure you don't want a `deepcopy()` here? (I'm not saying this is the case, but just to raise the point to make sure it's been considered.) :]

Comment (gtmanfred) [reply salt/cloud/__init__.py]: we are doing a copy on the list, which is just the keys, which can't be complex data structures, so we should be fine.

**Review by cachedout — APPROVED** [review]

File: salt/cloud/__init__.py (hunk 1 above)
Comment (techhat) [reply salt/cloud/__init__.py]: It's not a very complex data structure. I don't think we'll need a `deepcopy()` here.

**Pull request merged**